<h1>Guided Project: Clean and Analyze Employee Exit Surveys</h1>

In this guided project, we'll work with exit surveys from employees of the Department of Education, Training and Employment (DETE) and the Technical and Further Education (TAFE) institute in Queensland, Australia.

In this project, we'll play the role of data analyst and pretend our stakeholders want to know the following:

- Are employees who only worked for the institutes for a short period of time resigning due to some kind of dissatisfaction? What about employees who have been there longer?

- Are younger employees resigning due to some kind of dissatisfaction? What about older employees?


They want us to combine the results for both surveys to answer these questions. However, although both used the same survey template, one of them customized some of the answers. In the guided steps, we'll aim to do most of the data cleaning and get you started analyzing the first question.

In [1]:
import pandas as pd
import numpy as np
dete_survey = pd.read_csv('dete_survey.csv')
tafe_survey = pd.read_csv('tafe_survey.csv')

After reading in the files, let's now explore what these two datasets are about.

<h2>DETE Survey</h2>

In [2]:
dete_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 822 entries, 0 to 821
Data columns (total 56 columns):
 #   Column                               Non-Null Count  Dtype 
---  ------                               --------------  ----- 
 0   ID                                   822 non-null    int64 
 1   SeparationType                       822 non-null    object
 2   Cease Date                           822 non-null    object
 3   DETE Start Date                      822 non-null    object
 4   Role Start Date                      822 non-null    object
 5   Position                             817 non-null    object
 6   Classification                       455 non-null    object
 7   Region                               822 non-null    object
 8   Business Unit                        126 non-null    object
 9   Employment Status                    817 non-null    object
 10  Career move to public sector         822 non-null    bool  
 11  Career move to private sector        822 non-

In [3]:
dete_survey.head(3)

ID                    SeparationType Cease Date DETE Start Date  \
0   1             Ill Health Retirement    08/2012            1984   
1   2  Voluntary Early Retirement (VER)    08/2012      Not Stated   
2   3  Voluntary Early Retirement (VER)    05/2012            2011   

  Role Start Date         Position Classification          Region  \
0            2004   Public Servant        A01-A04  Central Office   
1      Not Stated   Public Servant        AO5-AO7  Central Office   
2            2011  Schools Officer            NaN  Central Office   

                       Business Unit    Employment Status  ...  Kept informed  \
0  Corporate Strategy and Peformance  Permanent Full-time  ...              N   
1  Corporate Strategy and Peformance  Permanent Full-time  ...              N   
2               Education Queensland  Permanent Full-time  ...              N   

   Wellness programs  Health & Safety  Gender          Age  Aboriginal  \
0                  N                N    Male        56-60         NaN   
1                  N                N    Male        56-60         NaN   
2                  N                N    Male  61 or older         NaN   

   Torres Strait  South Sea  Disability  NESB  
0            NaN        NaN         NaN   Yes  
1            NaN        NaN         NaN   NaN  
2            NaN        NaN         NaN   NaN  

[3 rows x 56 columns]

The DETE Survey has 822 entries/rows and 56 columns, most of which are `string` and `bool` types, and one `int` type. The following variables have very few values:

- `Business Unit`
- `Aboriginal`
- `Torres Strait`
- `South Sea`
- `Disability`
- `NESB`

<h2>TAFE Survey</h2>

In [4]:
tafe_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 72 columns):
 #   Column                                                                                                                                                         Non-Null Count  Dtype  
---  ------                                                                                                                                                         --------------  -----  
 0   Record ID                                                                                                                                                      702 non-null    float64
 1   Institute                                                                                                                                                      702 non-null    object 
 2   WorkArea                                                                                                                                  

In [5]:
tafe_survey.head(3)

Record ID                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   

                   WorkArea  CESSATION YEAR Reason for ceasing employment  \
0  Non-Delivery (corporate)          2010.0              Contract Expired   
1  Non-Delivery (corporate)          2010.0                    Retirement   
2       Delivery (teaching)          2010.0                    Retirement   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   
2                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    
2                                                  -    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   

   ...  \
0  ...   
1  ...   
2  ...   

  Workplace. Topic:Does your workplace promote a work culture free from all forms of unlawful discrimination?  \
0                                                Yes                                                            
1                                                Yes                                                            
2                                                Yes                                                            

  Workplace. Topic:Does your workplace promote and practice the principles of employment equity?  \
0                                                Yes                                               
1                                                Yes                                               
2                                                Yes                                               

  Workplace. Topic:Does your workplace value the diversity of its employees?  \
0                                                Yes                           
1                                                Yes                           
2                                                Yes                           

  Workplace. Topic:Would you recommend the Institute as an employer to others?  \
0                                                Yes                             
1                                                Yes                             
2                                                Yes                             

  Gender. What is your Gender? CurrentAge. Current Age  \
0                       Female                  26  30   
1                          NaN                     NaN   
2                          NaN                     NaN   

  Employment Type. Employment Type Classification. Classification  \
0              Temporary Full-time            Administration (AO)   
1                              NaN                            NaN   
2                              NaN                            NaN   

  LengthofServiceOverall. Overall Length of Service at Institute (in years)  \
0                                                1-2                          
1                                                NaN                          
2                                                NaN                          

  LengthofServiceCurrent. Length of Service at current workplace (in years)  
0                     

The TAFE Survey set has 702 entries and 72 columns/variables. There are many columns with null data of up to 270 entries or more than 38% of the dataset. All columns are string type except 2 of which are float type. This set will take more work to comb through.

Furthermore, we need to combine these datasets together and there are significant differences in the number of entries and the number of variables/columns. However, narrowing down our objective, we are only seeking information on the relationship between reason for resignation/separation. So we can focus on the following variables in each dataset:

<b>DETE Survey</b>
- `ID`: to identify the participant in the survey
- `SeparationType`: reason person's employment ended
- `Cease Date`: year or month person's employment ended
- `DETE Start Date`: year person began employment with DETE

<b>TAFE Survey</b>
- `Record ID`: ID used to identify participant in the survey
- `Reason for ceasing employment`: reason why person's employment ended
- `LengthofServiceOverall. Overall Length of Service at Institute (in years)`: length of employment (years)

Based on the information above let's look specifically at only the variables under consideration:

<h2>DETE Variables</h2>

In [6]:
# Select only columns under consideration and assign to new DF
dete_focus = dete_survey.iloc[:, 0:4]
dete_focus.isnull().sum()

ID                 0
SeparationType     0
Cease Date         0
DETE Start Date    0
dtype: int64

Fortunately there are no null values in our modified set. Let's look at the distribution of the values to ensure there are no values which exist but denote a 'null' status since these are `string` object types with the exception of `ID`.

In [7]:
dete_focus.head()

ID                    SeparationType Cease Date DETE Start Date
0   1             Ill Health Retirement    08/2012            1984
1   2  Voluntary Early Retirement (VER)    08/2012      Not Stated
2   3  Voluntary Early Retirement (VER)    05/2012            2011
3   4         Resignation-Other reasons    05/2012            2005
4   5                    Age Retirement    05/2012            1970

In [8]:
# create list for dete focus columns
dete_cols = dete_focus.columns

# apply value_counts
for col in dete_cols:
    print(dete_survey[col].value_counts())

823    1
270    1
280    1
279    1
278    1
      ..
546    1
545    1
544    1
543    1
1      1
Name: ID, Length: 822, dtype: int64
Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15
Name: SeparationType, dtype: int64
2012          344
2013          200
01/2014        43
12/2013        40
09/2013        34
Not Stated     34
06/2013        27
07/2013        22
10/2013        20
11/2013        16
08/2013        12
05/2013         7
05/2012         6
04/2014         2
04/2013         2
02/2014         2
07/2014         2
08/2012         2
09/2010         1
07/2006         1
2014            1
2010            1
09/2014         1
07/2012         1
11

Within the DETE Survey set, the `Cease Date` and `DETE Start Date` will require manipulation. The `Cease Date` variable has values in both month and year format and also has 34 'Not Stated' values. The `DETE Start Date` variable has all the same 'year' format but has 73 'Not Stated' values.

<h2>TAFE Variables</h2>

In [9]:
# Select only columns under consideration and assign to new DF
tafe_focus = tafe_survey.iloc[:, [0, 4, 70]]
tafe_focus.isnull().sum()

Record ID                                                                      0
Reason for ceasing employment                                                  1
LengthofServiceOverall. Overall Length of Service at Institute (in years)    106
dtype: int64

We have 1 null value in `Reason for ceasing employment` and 106 null values in the `LengthofService...` variable.

In [10]:
# create list for dete focus columns
tafe_cols = tafe_focus.columns

# apply value_counts
for col in tafe_cols:
    print(tafe_survey[col].value_counts())

6.342194e+17    1
6.349929e+17    1
6.342081e+17    1
6.345951e+17    1
6.341719e+17    1
               ..
6.348914e+17    1
6.342080e+17    1
6.349520e+17    1
6.342684e+17    1
6.341820e+17    1
Name: Record ID, Length: 702, dtype: int64
Resignation                 340
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23
Name: Reason for ceasing employment, dtype: int64
Less than 1 year      147
1-2                   102
3-4                    96
11-20                  89
More than 20 years     71
5-6                    48
7-10                   43
Name: LengthofServiceOverall. Overall Length of Service at Institute (in years), dtype: int64


The first two columns in TAFE appear clean and the third column also appears clean but different than the DETE set as this variable for length of service is not a date, like a year or month, but a categorical variable for length of employment.

We will need to take this into consideration when we combine these datasets.

<h2>Cleaning the Data</h2>

We can first make the following observations:

- The dete_survey dataframe contains 'Not Stated' values that indicate values are missing, but they aren't represented as NaN.
- Both the dete_survey and tafe_survey dataframes contain many columns that we don't need to complete our analysis.
- Each dataframe contains many of the same columns, but the column names are different.
- There are multiple columns/answers that indicate an employee resigned because they were dissatisfied.

To start, we'll handle the first two issues.

In [11]:
# read DETE file in again substituting 'NaN' for 'Not Stated' values.
dete_survey = pd.read_csv('dete_survey.csv', na_values='Not Stated').copy()

#pull columns to remove then remove and assign to new var
cols_to_remove = dete_survey.columns[28:49]
dete_survey_updated = dete_survey.drop(labels=cols_to_remove, axis=1).copy()

#pull columns to remove then remove and assign to new var
cols_to_remove = tafe_survey.columns[17:66]
tafe_survey_updated = tafe_survey.drop(labels=cols_to_remove, axis=1)

print(dete_survey_updated.info())
print(tafe_survey_updated.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 822 entries, 0 to 821
Data columns (total 35 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   ID                                   822 non-null    int64  
 1   SeparationType                       822 non-null    object 
 2   Cease Date                           788 non-null    object 
 3   DETE Start Date                      749 non-null    float64
 4   Role Start Date                      724 non-null    float64
 5   Position                             817 non-null    object 
 6   Classification                       455 non-null    object 
 7   Region                               717 non-null    object 
 8   Business Unit                        126 non-null    object 
 9   Employment Status                    817 non-null    object 
 10  Career move to public sector         822 non-null    bool   
 11  Career move to private sector   

Above you can see that unnecessary columns have been removed.

Next, let's turn our attention to the column names. Each dataframe contains many of the same columns, but the column names are different. 

<h3>Standardize Column Names</h3>

In [12]:
# Rename remaining columns in `dete_survey_updated`
dete_survey_updated.columns = (dete_survey_updated.columns.str.replace(' ', '_')
                               .str.strip().str.lower())
dete_survey_updated.columns

Index(['id', 'separationtype', 'cease_date', 'dete_start_date',
       'role_start_date', 'position', 'classification', 'region',
       'business_unit', 'employment_status', 'career_move_to_public_sector',
       'career_move_to_private_sector', 'interpersonal_conflicts',
       'job_dissatisfaction', 'dissatisfaction_with_the_department',
       'physical_work_environment', 'lack_of_recognition',
       'lack_of_job_security', 'work_location', 'employment_conditions',
       'maternity/family', 'relocation', 'study/travel', 'ill_health',
       'traumatic_incident', 'work_life_balance', 'workload',
       'none_of_the_above', 'gender', 'age', 'aboriginal', 'torres_strait',
       'south_sea', 'disability', 'nesb'],
      dtype='object')

In [13]:
# Rename remaining columns in `tafe_survey_updated`
dict = {
    'Record ID': 'id',
    'CESSATION YEAR': 'cease_date',
    'Reason for ceasing employment': 'separationtype',
    'Gender. What is your Gender?': 'gender',
    'CurrentAge. Current Age': 'age',
    'Employment Type. Employment Type': 'employment_status',
    'Classification. Classification': 'position',
    'LengthofServiceOverall. Overall Length of Service at Institute (in years)': 'institute_service',
    'LengthofServiceCurrent. Length of Service at current workplace (in years)': 'role_service'
}


tafe_survey_updated.rename(columns=dict, inplace=True)
 
tafe_survey_updated.columns

Index(['id', 'Institute', 'WorkArea', 'cease_date', 'separationtype',
       'Contributing Factors. Career Move - Public Sector ',
       'Contributing Factors. Career Move - Private Sector ',
       'Contributing Factors. Career Move - Self-employment',
       'Contributing Factors. Ill Health',
       'Contributing Factors. Maternity/Family',
       'Contributing Factors. Dissatisfaction',
       'Contributing Factors. Job Dissatisfaction',
       'Contributing Factors. Interpersonal Conflict',
       'Contributing Factors. Study', 'Contributing Factors. Travel',
       'Contributing Factors. Other', 'Contributing Factors. NONE', 'gender',
       'age', 'employment_status', 'position', 'institute_service',
       'role_service'],
      dtype='object')

We made changes to the column names above for uniformity, simplicity, and to make it more intuitive when combining the tables.

We renamed the columns that we'll use in our analysis. Next, let's remove more of the data we don't need.

Recall that our end goal is to answer the following question:

- Are employees who have only worked for the institutes for a short period of time resigning due to some kind of dissatisfaction? 
- What about employees who have been at the job longer?

If we look at the unique values in the `separationtype` columns in each dataframe, we'll see that each contains a couple of different separation types. For this project, we'll only analyze survey respondents who resigned, so their separation type contains the string `'Resignation'`.

In [14]:
dete_survey_updated['separationtype'].value_counts()

Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15
Name: separationtype, dtype: int64

In [15]:
tafe_survey_updated['separationtype'].value_counts()

Resignation                 340
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23
Name: separationtype, dtype: int64

In each data series, we are only looking for `'Resignation'` type values.

In [16]:
dete_resignations = dete_survey_updated['separationtype'].str.startswith('Resignation').copy()
dete_resignations

0      False
1      False
2      False
3       True
4      False
       ...  
817    False
818    False
819     True
820    False
821     True
Name: separationtype, Length: 822, dtype: bool

In [17]:
tafe_resignations = tafe_survey_updated['separationtype'] == 'Resignation'
tafe_resignations

0      False
1      False
2      False
3       True
4       True
       ...  
697     True
698     True
699     True
700    False
701     True
Name: separationtype, Length: 702, dtype: bool

So above we have our boolean indexes to filter for only resignations.

Now, before we start cleaning and manipulating the rest of our data, let's verify that the data doesn't contain any major inconsistencies (to the best of our knowledge). When you're working with real world data, don't assume that the data you're analyzing isn't corrupted in some way!

It may not always be possible to catch all of these errors, but by making sure the data seems reasonable to the best of our knowledge, we can stop ourselves from completing a data analysis project that winds up being useless because of bad data.

In this step, we'll focus on verifying that the years in the cease_date and dete_start_date columns make sense.

In [18]:
dete_resignations = dete_survey_updated[dete_resignations].copy()
dete_resignations['cease_date'].value_counts()

2012       126
2013        74
01/2014     22
12/2013     17
06/2013     14
09/2013     11
07/2013      9
11/2013      9
10/2013      6
08/2013      4
05/2013      2
05/2012      2
2010         1
09/2010      1
07/2012      1
07/2006      1
Name: cease_date, dtype: int64

In [19]:
pattern = r"(?P<month>[0-9][0-9][/])?(?P<year>[0-9][0-9][0-9][0-9])"    
split = dete_resignations['cease_date'].str.extract(pattern).copy()
year = split['year'].astype(float).copy()
dete_resignations.loc[:, ['cease_date']] = pd.Series(data=year)
dete_resignations['cease_date'].value_counts()

2013.0    146
2012.0    129
2014.0     22
2010.0      2
2006.0      1
Name: cease_date, dtype: int64

In [20]:
dete_resignations['dete_start_date'].value_counts()

2011.0    24
2008.0    22
2007.0    21
2012.0    21
2010.0    17
2005.0    15
2004.0    14
2009.0    13
2006.0    13
2013.0    10
2000.0     9
1999.0     8
1996.0     6
2002.0     6
1992.0     6
1998.0     6
2003.0     6
1994.0     6
1993.0     5
1990.0     5
1980.0     5
1997.0     5
1991.0     4
1989.0     4
1988.0     4
1995.0     4
2001.0     3
1985.0     3
1986.0     3
1983.0     2
1976.0     2
1974.0     2
1971.0     1
1972.0     1
1984.0     1
1982.0     1
1987.0     1
1975.0     1
1973.0     1
1977.0     1
1963.0     1
Name: dete_start_date, dtype: int64

Above, we extracted the year from the string object in the `cease_date` column, reinserting only the year (no month) as float to make it consistent.

In [21]:
tafe_resignations = tafe_survey_updated[tafe_resignations].copy()
tafe_resignations['cease_date'].value_counts()

2011.0    116
2012.0     94
2010.0     68
2013.0     55
2009.0      2
Name: cease_date, dtype: int64

Above, the dates for `cease_date` in tafe_resignations appears correct.

<h3>Create New Column</h3>

Since the `dete_resignations` set has not `'institute_service'` variable/column like that of `tafe_resignations` we will need to create one.

In [22]:
dete_resignations['institute_service'] = (dete_resignations['cease_date'] 
                                          - dete_resignations['dete_start_date'])

In [23]:
dete_resignations['institute_service'].value_counts()

5.0     23
1.0     22
3.0     20
0.0     20
6.0     17
4.0     16
9.0     14
2.0     14
7.0     13
13.0     8
8.0      8
20.0     7
15.0     7
10.0     6
22.0     6
14.0     6
17.0     6
12.0     6
16.0     5
18.0     5
23.0     4
11.0     4
24.0     4
39.0     3
19.0     3
21.0     3
32.0     3
28.0     2
26.0     2
25.0     2
30.0     2
36.0     2
29.0     1
33.0     1
42.0     1
27.0     1
41.0     1
35.0     1
38.0     1
34.0     1
49.0     1
31.0     1
Name: institute_service, dtype: int64

Above we created an `'institute_service'` column and we can see the results look reasonable (in years) based on the `value_counts` results.

<h3>Identify Dissatisfied Employees</h3>

Next, we'll identify any employees who resigned because they were dissatisfied. Below are the columns we'll use to categorize employees as "dissatisfied" from each dataframe.

1. tafe_survey_updated:
- Contributing Factors. Dissatisfaction
- Contributing Factors. Job Dissatisfaction

2. dete_survey_updated:
- job_dissatisfaction
- dissatisfaction_with_the_department
- physical_work_environment
- lack_of_recognition
- lack_of_job_security
- work_location
- employment_conditions
- work_life_balance
- workload

If the employee indicated any of the factors above caused them to resign, we'll mark them as `dissatisfied` in a new column.

In [24]:
tafe_resignations['Contributing Factors. Dissatisfaction'].value_counts()

-                                         277
Contributing Factors. Dissatisfaction      55
Name: Contributing Factors. Dissatisfaction, dtype: int64

In [25]:
tafe_resignations['Contributing Factors. Job Dissatisfaction'].value_counts()

-                      270
Job Dissatisfaction     62
Name: Contributing Factors. Job Dissatisfaction, dtype: int64

In [26]:
# function to relabel columns as True or False
def relabel(elem):
    elem = str(elem)
    if 'Dissatisfaction' in elem:
        return True
    elif elem == '-': 
        return False
    elif elem == 'NaN' or elem == np.nan:
        return np.nan
    
columns = ['Contributing Factors. Dissatisfaction', 'Contributing Factors. Job Dissatisfaction']
dissat = tafe_resignations[columns].applymap(relabel)
dissat['Contributing Factors. Dissatisfaction'].value_counts()

False    277
True      55
Name: Contributing Factors. Dissatisfaction, dtype: int64

In [27]:
dissat['Contributing Factors. Job Dissatisfaction'].value_counts()

False    270
True      62
Name: Contributing Factors. Job Dissatisfaction, dtype: int64

If any of the columns listed above contain a `True` value, we'll add a `True` value to a new column named `'dissatisfied'`.

In [28]:
tafe_resignations['dissatisfied'] = dissat.any(axis=1, skipna=False).copy()
tafe_resignations['dissatisfied'].value_counts()

False    241
True      91
Name: dissatisfied, dtype: int64

As a result we see that there are 91 resignations in the `tafe_survey` due to dissatisfaction.

Now we run `df.any()` on the `'dete_resignations'` set.

In [29]:
# Collect columns related to job satisfaction
dete_cols = dete_resignations.columns[13:20]
dete_cols2 = dete_resignations.columns[25:27]

# Append list 2 to list 1 to combine columns
dete_cols = dete_cols.append(dete_cols2)
dete_cols

Index(['job_dissatisfaction', 'dissatisfaction_with_the_department',
       'physical_work_environment', 'lack_of_recognition',
       'lack_of_job_security', 'work_location', 'employment_conditions',
       'work_life_balance', 'workload'],
      dtype='object')

In [30]:
# Create dataframe with columns for truth table
dete_truth = dete_resignations[dete_cols].copy()
dete_truth.head()

job_dissatisfaction  dissatisfaction_with_the_department  \
3                 False                                False   
5                 False                                False   
8                 False                                False   
9                  True                                 True   
11                False                                False   

    physical_work_environment  lack_of_recognition  lack_of_job_security  \
3                       False                False                 False   
5                       False                False                 False   
8                       False                False                 False   
9                       False                False                 False   
11                      False                False                 False   

    work_location  employment_conditions  work_life_balance  workload  
3           False                  False              False     False  
5           False                   True              False     False  
8           False                  False              False     False  
9           False                  False              False     False  
11          False                  False              False     False

In [31]:
dete_truth['dissatisfied'] = dete_truth.any(axis=1, skipna=False).copy()
dete_truth['dissatisfied'].value_counts()

False    162
True     149
Name: dissatisfied, dtype: int64

Now we add the `dissatisfied` column to our `dete_resignations` dataframe. Then we check to ensure we have our True/False `dissatisfied` columns.

In [32]:
dete_resignations['dissatisfied'] = dete_truth['dissatisfied']

In [33]:
# Create copy of results
dete_resignations_up = dete_resignations.copy()
tafe_resignations_up = tafe_resignations.copy()
print(dete_resignations_up.head())
print(tafe_resignations_up.head())

    id                        separationtype  cease_date  dete_start_date  \
3    4             Resignation-Other reasons      2012.0           2005.0   
5    6             Resignation-Other reasons      2012.0           1994.0   
8    9             Resignation-Other reasons      2012.0           2009.0   
9   10            Resignation-Other employer      2012.0           1997.0   
11  12  Resignation-Move overseas/interstate      2012.0           2009.0   

    role_start_date          position classification                region  \
3            2006.0           Teacher        Primary    Central Queensland   
5            1997.0  Guidance Officer            NaN        Central Office   
8            2009.0           Teacher      Secondary      North Queensland   
9            2008.0      Teacher Aide            NaN                   NaN   
11           2009.0           Teacher      Secondary  Far North Queensland   

           business_unit    employment_status  ...  none_of_the_abov

<h2>Combining the Datasets</h2>

Now, we're finally ready to combine our datasets! Our end goal is to aggregate the data according to the `institute_service` column.

In [34]:
dete_resignations_up['institute'] = 'DETE'
tafe_resignations_up['institute'] = 'TAFE'

combined = pd.concat([dete_resignations_up, tafe_resignations_up], ignore_index=True)

In [35]:
combined['dissatisfied'].value_counts()

False    403
True     240
Name: dissatisfied, dtype: int64

In [36]:
combined_updated = combined.dropna(axis=1, thresh=500).copy()
combined_updated.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 651 entries, 0 to 650
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 651 non-null    float64
 1   separationtype     651 non-null    object 
 2   cease_date         635 non-null    float64
 3   position           598 non-null    object 
 4   employment_status  597 non-null    object 
 5   gender             592 non-null    object 
 6   age                596 non-null    object 
 7   institute_service  563 non-null    object 
 8   dissatisfied       643 non-null    object 
 9   institute          651 non-null    object 
dtypes: float64(2), object(8)
memory usage: 51.0+ KB


Above we combined the datasets using an outer join with the `concat()` function. Then, we removed any columns in the resulting dataframe which have less than 500 non-null values.

<h3>Clean the Service Column</h3>

As you can see below, we have various formats and overlaps of values in the `institute_service` variable. Let's clean this up. We will discretize it based on the <a href='https://www.businesswire.com/news/home/20171108006002/en/Age-Number-Engage-Employees-Career-Stage'>categories in this article</a>, which makes the argument that understanding employee's needs according to career stage instead of age is more effective.

We'll use the slightly modified definitions below:

- New: Less than 3 years at a company
- Experienced: 3-6 years at a company
- Established: 7-10 years at a company
- Veteran: 11 or more years at a company

In [37]:
combined_updated['institute_service'].value_counts()

Less than 1 year      73
1-2                   64
3-4                   63
5-6                   33
11-20                 26
5.0                   23
1.0                   22
7-10                  21
3.0                   20
0.0                   20
6.0                   17
4.0                   16
9.0                   14
2.0                   14
7.0                   13
More than 20 years    10
13.0                   8
8.0                    8
20.0                   7
15.0                   7
14.0                   6
17.0                   6
12.0                   6
10.0                   6
22.0                   6
18.0                   5
16.0                   5
24.0                   4
23.0                   4
11.0                   4
39.0                   3
19.0                   3
21.0                   3
32.0                   3
36.0                   2
25.0                   2
26.0                   2
28.0                   2
30.0                   2
42.0                   1


The values above are in string format so we will use vectorized string operations to convert the discretized ones to single number. Thereafter we will map the values back into our new discretized formats.

In [38]:
# RegEx match 1 or 2 consecutive digits in each value
pattern = r"([0-9]{1,2})"
years = combined_updated['institute_service'].str.extract(pattern)

# reset index on dataframe to prevent `SettingwithCopyWarning` since we filtered rows
combined_updated.reset_index(drop=True, inplace=True)

# updated column
combined_updated['institute_service'] = years
 
# check that `institute_service` updated
print(combined_updated.tail())

               id separationtype  cease_date                 position  \
646  6.350660e+17    Resignation      2013.0         Operational (OO)   
647  6.350668e+17    Resignation      2013.0  Teacher (including LVT)   
648  6.350677e+17    Resignation      2013.0                      NaN   
649  6.350704e+17    Resignation      2013.0  Teacher (including LVT)   
650  6.350730e+17    Resignation      2013.0      Administration (AO)   

       employment_status  gender     age institute_service dissatisfied  \
646  Temporary Full-time    Male  21  25                 5        False   
647  Temporary Full-time    Male   51-55                 1        False   
648                  NaN     NaN     NaN               NaN        False   
649  Permanent Full-time  Female   51-55                 5        False   
650      Contract/casual  Female  26  30                 3        False   

    institute  
646      TAFE  
647      TAFE  
648      TAFE  
649      TAFE  
650      TAFE  


In [39]:
combined_updated['institute_service'].astype('float')
combined_updated['institute_service'].value_counts()

1     137
3      63
5      33
11     26
7      21
20     10
Name: institute_service, dtype: int64

Now above we have only numeric years in `institute_service` in `float` format.

Next, we map each value to one of the career stages.

In [44]:
def map_year(val):
    val = float(val)
    if pd.isnull(val):
        return "NA"
    elif val < 3:
        return "New"
    elif (val > 3) and (val <= 6):
        return "Experienced"
    elif (val > 6) and (val <= 10):
        return "Established"
    elif val > 10:
        return "Veteran"
    
combined_updated['service_cat'] = combined_updated['institute_service'].apply(map_year)
combined_updated['service_cat'].value_counts()

NA             361
New            137
Veteran         36
Experienced     33
Established     21
Name: service_cat, dtype: int64

Above you see that we have discretized the years of service into categories.

<h3>Perform Initial Analysis of `dissatisfied` Workers</h3>

In [41]:
combined_updated['dissatisfied'].value_counts(dropna=False)

False    403
True     240
NaN        8
Name: dissatisfied, dtype: int64

In [42]:
combined_updated['dissatisfied'] = combined_updated['dissatisfied'].fillna(False)
combined_updated['dissatisfied'].astype('bool')
combined_updated['dissatisfied'].value_counts(dropna=False)

False    411
True     240
Name: dissatisfied, dtype: int64

We converted the `NaN` values to `False` values.

In [48]:
dissatisfied = combined_updated.pivot_table(values='dissatisfied', index='service_cat', margins=True)
dissatisfied

dissatisfied
service_cat              
Established      0.333333
Experienced      0.272727
NA               0.451524
New              0.262774
Veteran          0.277778
All              0.382653

Above we aggregated the categories and calculated the mean percentage dissatisfied for each category. Overall, the average percentage of employees resigned due to dissatisfaction is 38.27%. Established workers were slightly more likely to resign due to dissatisfaction compared to the others - `Experience`, `New`, and `Veteran` employees. The fact that we have an `NA` category has skewed the mean up higher since more than 45% of the uncategorized employees resigned due to dissatisfaction. We were unable to determine the length of service for this category but kept the data in. 